### PR TITLE
Release IPs on bad IPAM result

### LINKED ISSUE #325: CNI plugin can leak IP addresses
Seen on Calico CNI v1.8.0
When the calico CNI plugin calls an IPAM plugin, it can end up returning an error to the runtime even though the IPAM plugin didn't return an error. For this to be OK, the networking plugin needs to ensure that it is instructing the IPAM plugin to cleanup, otherwise addresses can be leaked.

e.g. in this code, at least the last two return statements (and maybe the first one too) should be calling ipam.ExecDel.
```
		case ipAddrs == "" && ipAddrsNoIpam == "":
			// Call IPAM plugin if ipAddrsNoIpam or ipAddrs annotation is not present.
			logger.Debugf("Calling IPAM plugin %s", conf.IPAM.Type)
			ipamResult, err := ipam.ExecAdd(conf.IPAM.Type, args.StdinData)
			if err != nil {
				return nil, err
			}
			logger.Debugf("IPAM plugin returned: %+v", ipamResult)

			// Convert IPAM result into current Result.
			// IPAM result has a bunch of fields that are optional for an IPAM plugin
			// but required for a CNI plugin, so this is to populate those fields.
			// See CNI Spec doc for more details.
			result, err = current.NewResultFromResult(ipamResult)
			if err != nil {
				return nil, err
			}

			if len(result.IPs) == 0 {
				return nil, errors.New("IPAM plugin returned missing IP config")
			}
```

### PR DESCRIPTION
**Pull request by caseydavenport**
Fixes #325

### CODE REVIEW

**Comment by gunjan5** [comment]
@caseydavenport I think we also need to do this at https://github.com/projectcalico/cni-plugin/blob/master/calico.go#L157 and https://github.com/projectcalico/cni-plugin/blob/master/calico.go#L162

File: calico_cni_k8s_test.go [hunk 1]
```diff
@@ -885,6 +885,59 @@ var _ = Describe("CalicoCni", func() {
 					_, err = DeleteContainer(netconf, netnspath, "")
 					Expect(err).ShouldNot(HaveOccurred())
 				})
+
+				Context("when IPAM returns an empty result", func() {
```
Comment (caseydavenport) [inline calico_cni_k8s_test.go]: This isn't done yet, didn't mean to add this to the PR yet :)

File: Makefile [hunk 2]
```diff
@@ -4,7 +4,7 @@
 
 SRCFILES=calico.go $(wildcard utils/*.go) $(wildcard k8s/*.go) ipam/calico-ipam.go
 TEST_SRCFILES=$(wildcard test_utils/*.go) $(wildcard calico_cni_*.go)
-LOCAL_IP_ENV?=$(shell ip route get 8.8.8.8 | head -1 | awk '{print $$7}')
+LOCAL_IP_ENV?=$(shell docker run -ti --rm --net=host busybox ip route get 8.8.8.8 | head -1 | awk '{print $$7}')
```
Comment (caseydavenport) [inline Makefile]: Just so this now works on mac as well :)

File: calico_cni_k8s_test.go [hunk 3]
```diff
@@ -885,6 +885,55 @@ var _ = Describe("CalicoCni", func() {
...
+							Spec: v1.PodSpec{Containers: []v1.Container{{
+								Name:  fmt.Sprintf("container-%s", name),
+								Image: "ignore",
+							}}},
+						})
+						Expect(err).NotTo(HaveOccurred())
+
+						logger.Infof("Created POD object: %v", pod)
+
+						_, netnspath, _, _, _, _, _, err := CreateContainer(netConf, name, "")
```
Comment (caseydavenport) [inline calico_cni_k8s_test.go]: So this seems to be working (can tell from the logs), but I don't know how to assert that the IPs have been actually released by the plugin.

I could update `echo-ipam` to do some fake IP allocation / releasing perhaps using files on disk and then assert that the file is removed in the test?

@tomdee

File: Makefile (hunk 2 above)
Comment (tomdee) [inline Makefile]: Lol, nice workaround.

File: Makefile [hunk 4]
```diff
@@ -109,6 +109,9 @@ $(DEPLOY_CONTAINER_MARKER): Dockerfile build-containerized fetch-cni-bins
 .PHONY: fetch-cni-bins
 fetch-cni-bins: dist/flannel dist/loopback dist/host-local
 
+dist/echo-ipam:
```
Comment (tomdee) [inline Makefile]: Why not just store echo-ipam in dist?

File: calico.go [hunk 5]
```diff
@@ -156,10 +156,12 @@ func cmdAdd(args *skel.CmdArgs) error {
 			// See CNI Spec doc for more details.
 			result, err = current.NewResultFromResult(ipamResult)
 			if err != nil {
+				ReleaseIPAllocation(logger, conf.IPAM.Type, args.StdinData)
```
Comment (tomdee) [inline calico.go]: How much thought have you put into checking all the places where this should be called?

Comment (tomdee) [inline calico.go]: And this probably needs test coverage too

File: calico_cni_k8s_test.go [hunk 6]
```diff
@@ -885,6 +885,55 @@ var _ = Describe("CalicoCni", func() {
...
+				Context("when IPAM returns an empty result", func() {
+					It("should try to release any IP addresses", func() {
+						netConf := fmt.Sprintf(`
+				{
+			      "cniVersion": "%s",
+				  "name": "net2",
+				  "type": "calico",
+				  "etcd_endpoints": "http://%s:2379",
+			 	  "ipam": {
+			    	 "type": "echo-ipam",
```
Comment (tomdee) [inline calico_cni_k8s_test.go]: formatting (indent)

File: calico_cni_k8s_test.go [hunk 7]
```diff
@@ -885,6 +885,55 @@ var _ = Describe("CalicoCni", func() {
...
+				 "return": "{}"
```
Comment (tomdee) [inline calico_cni_k8s_test.go]: I don't really get this line, is it left over from a previous design of your echo plugin?

Comment (caseydavenport) [inline calico_cni_k8s_test.go]: It is indeed.

File: Makefile (hunk 4 above)
Comment (caseydavenport) [inline Makefile]: `dist` isn't checked in at the moment.  I could do that if so desired though.  LMK.

**Comment by robbrockbank** [comment]
@caseydavenport :  should this be closed?

**Pull request closed without merging**